arch/arc/core/cache.c

```diff
@@ -110,36 +94,37 @@ static void dcache_flush_mlines(u32_t start_addr, u32_t size)
 				break;
 			}
 		} while (1);
-		start_addr += DCACHE_LINE_SIZE;
-	} while (start_addr <= end_addr);
+		start_addr = (void *)((uintptr_t)start_addr + DCACHE_LINE_SIZE);
+	} while ((uintptr_t)start_addr < end_addr);
 
 	arch_irq_unlock(key); /* --exit critical section-- */
 
 }
 
+void arch_dcache_invd(void *start_addr, size_t size)
+{
+	uintptr_t end_addr;
+	unsigned int key;
 
-/**
- *
- * @brief Flush d-cache lines to main memory
- *
- * No alignment is required for either <virt> or <size>, but since
- * sys_cache_flush() iterates on the d-cache lines, a d-cache line alignment for
- * both is optimal.
- *
- * The d-cache line size is specified either via the CONFIG_CACHE_LINE_SIZE
- * kconfig option or it is detected at runtime.
- *
- * @param start_addr the pointer to start the multi-line flush
- * @param size the number of bytes that are to be flushed
- *
- * @return N/A
- */
+	if (!dcache_available() || (size == 0U)) {
+		printk("Warn: cache not available or %d size\n", size);
+		return;
+	}
+	end_addr = (uintptr_t)start_addr + size;
```

Comment:
style: consider a local uintptr_t variable for the start address, rather than casting the arg in lots of places.

Reply:
Addressed with local variable start_addr_temp. thanks